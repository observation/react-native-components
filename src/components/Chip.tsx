import React, { useState } from 'react'
import {
  LayoutChangeEvent,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native'

import appTextStyle from '../styles/text'
import theme from '../styles/theme'

type Props = {
  text?: string
  textStyle?: StyleProp<TextStyle>
  containerStyle?: StyleProp<ViewStyle>
  onPress?: () => void
  disabled?: boolean
}

const Chip = ({ text, textStyle, containerStyle, onPress, disabled }: Props) => {
  const [borderRadius, setBorderRadius] = useState(theme.margin.common)
  return (
    <TouchableOpacity onPress={disabled ? undefined : onPress} activeOpacity={0.5} disabled={disabled}>
      <View
        onLayout={(event: LayoutChangeEvent) => setBorderRadius(event.nativeEvent.layout.height / 2)}
        style={[styles.chipContainer, { borderRadius }, containerStyle]}
      >
        <View style={styles.chipTextContainer}>
          <Text style={[styles.chipText, textStyle]}>{text}</Text>
        </View>
      </View>
    </TouchableOpacity>
  )
}

const styles = StyleSheet.create({
  chipTextContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
  },
  chipText: {
    ...appTextStyle.body,
    color: theme.color.white,
    lineHeight: theme.margin.common,
    fontSize: theme.font.size.medium,
  },
  chipContainer: {
    backgroundColor: theme.color.successLight,
    paddingHorizontal: theme.margin.common,
    paddingVertical: theme.margin.half,
    borderRadius: theme.margin.common,
    minWidth: 44,
    justifyContent: 'center',
  },
})

export default Chip
