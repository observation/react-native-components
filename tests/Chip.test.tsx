import React from 'react'

import { fireEvent, render } from '@testing-library/react-native'

import Chip from '../src/components/Chip'

describe('Chip', () => {
  describe('Rendering', () => {
    test('Required props', () => {
      // WHEN
      const { toJSON } = render(<Chip />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('All props', () => {
      // WHEN
      const { toJSON } = render(
        <Chip
          text={'some text'}
          onPress={jest.fn()}
          containerStyle={{ flex: 1 }}
          textStyle={{ fontSize: 1 }}
          disabled={true}
        />,
      )

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })
  })

  test('Handle press', () => {
    // GIVE
    const onPress = jest.fn()
    const { getByText } = render(<Chip text={'some text'} onPress={onPress} />)

    // WHEN
    fireEvent.press(getByText('some text'))

    // THEN
    expect(onPress).toHaveBeenCalled()
  })

  test('Handle press should not cause error when no press is set', () => {
    // GIVE
    const onPress = jest.fn()
    const { getByText } = render(<Chip text={'some text'} onPress={onPress} disabled={true} />)

    // WHEN
    fireEvent.press(getByText('some text'))

    // THEN
    expect(onPress).not.toHaveBeenCalled()
  })
})
