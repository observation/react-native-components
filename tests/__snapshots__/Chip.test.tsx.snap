// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip Rendering All props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "#9DC556",
          "borderRadius": 16,
          "justifyContent": "center",
          "minWidth": 44,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        {
          "borderRadius": 16,
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 16,
            },
            {
              "fontSize": 1,
            },
          ]
        }
      >
        some text
      </Text>
    </View>
  </View>
</View>
`;

exports[`Chip Rendering Required props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "#9DC556",
          "borderRadius": 16,
          "justifyContent": "center",
          "minWidth": 44,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        {
          "borderRadius": 16,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 16,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;
